core/src/banking_stage.rs

```diff
@@ -912,7 +923,9 @@ impl BankingStage {
                 // Take metrics action after forwarding packets
                 slot_metrics_tracker.apply_action(metrics_action);
             }
-            _ => (),
+            BufferedPacketsDecision::Hold => {
+                slot_metrics_tracker.apply_action(metrics_action);
```

Comment:
This does nothing, right?

Reply:
actually, it does something meaning. this stands by its own as another small pr.. I'll explain there.

Reply:
after rabbit hole digging, i've decided to remove this code altogether: 567e2a2327f68f28e5a81c231c28dff40a95e350

the bottom of this story is that `SlotMetricsTracker` contains a semantic ambiguity. namely, it isn't properly handling a measurable gap between each `BankStart`s between the 4 slots of a single leader's round.

let me explain the unhappy field trip with a wall of text...

Firstly, I intentionally started to call `.apply_action()` here to emit banking tracer's Bank(Ended) events _only once_ per thread.

Without this `.apply_action()`, it emitted tons of Bank(Ended) events for each slot/banking thread (def. not desirable). 

That's because its code is placed inside `slot_metrics_tracker.check_leader_slot_boundary()`. Albeit its name, i thought it's _not idempotent_ after code read, because there's already precedent there: `.mark_slot_end_detected()`. so, i added code there instead of emitting in `.apply_action()` for timing accuracy, hoping the code path is touched only once for each state transition under caller's responsibility.

However, it turned out false wish. `ReportAndResetTracker` from `check_leader_slot_boundary()` _isn't consumed_ if there's no `BankStart` at the moment (this `BufferedPacketsDecision::Hold` match arm).  That means `check_leader_slot_boundary()` _keeps_ to return `ReportAndResetTracker`. This happens quite casully when freezing a prev bank/creating a new bank to set it to `PohRecorder`. So we're also repeatedly rewriting `.mark_slot_end_detected()` now in master branch too. (I think it's minor problem...)

So, the original problem of too many `Bank(Ended)` was fixed by properly consuming that event here.

Then, it popped up different issue: SlotMetricsTracker started to discard metrics until the new bank is set to `PohRecorder`. This especially affects `UnprocessedTransactionStorage`-related metrics. 

Come to think of it, i think these discarded metrics should be counted towards _next slot_'s `Metrics` as the current slot already reached to poh max height. after all, they're buffered to be processed _in the next slot_. however, currently it's counted towards to the current slot.  So, current metrics is a bit skewed to contain some extra numbers after `BankStart` got `None`.

Also, I can't simply move the Bank(Ended) emitting code into `.apply_action()`. its timing almost same as next slot's `Bank(Started)` without this `.apply_action()`. but then, i'd like to avoid to introduce the above semantics change.

now, i was starting to think it's too bothersome to fix them all, just for non-essential feature of this pr. so, i reverted.

Lastly, i also spotted another issue: `BufferedPacketsDecision::Forward` is emitted wrongly due to race condition. I think we're throwing away precious buffered txes into forwarding to some extent (not dug in depth, yet)

...phew :)

Reply:
> Lastly, i also spotted another issue: BufferedPacketsDecision::Forward is emitted wrongly due to race condition. I think we're throwing away precious buffered txes into forwarding to some extent at the duration of these bank freeze gaps (not dug in depth, yet)

This is something we should dig more in to. I thought during the gaps we'd just be emitting `Hold`, at least that's what i've observed in my tests. I will also need to take a look at that code (now nicely separated into `decision_maker.rs` 😄 )